// @bazel/typescript/internal/tsetse/util/allowlist.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Allowlist = exports.ExemptionReason = void 0;
/**
 * The categories of exemption entries.
 */
var ExemptionReason;
(function (ExemptionReason) {
    /** No reason. */
    ExemptionReason[ExemptionReason["UNSPECIFIED"] = 0] = "UNSPECIFIED";
    /** Code that has to be grandfathered in (no guarantees). */
    ExemptionReason[ExemptionReason["LEGACY"] = 1] = "LEGACY";
    /**
     * Code that does not enter the scope of this particular check  (no
     * guarantees).
     */
    ExemptionReason[ExemptionReason["OUT_OF_SCOPE"] = 2] = "OUT_OF_SCOPE";
    /** Manually reviewed exceptions (supposedly okay). */
    ExemptionReason[ExemptionReason["MANUALLY_REVIEWED"] = 3] = "MANUALLY_REVIEWED";
})(ExemptionReason = exports.ExemptionReason || (exports.ExemptionReason = {}));
/**
 * A complete allowlist with all related AllowlistEntry grouped together, with
 * ExemptionReason ignored since it is purely for documentary purposes.
 */
class Allowlist {
    constructor(allowlistEntries) {
        this.allowlistedPrefixes = [];
        this.allowlistedRegExps = [];
        // To avoid repeated computation for allowlisting queries with the same file
        // path, create a memoizer to cache known results. This is useful in watch
        // mode (and possible in language service) when the same files can be compiled
        // repeatedly.
        this.allowlistMemoizer = new Map();
        if (allowlistEntries) {
            for (const e of allowlistEntries) {
                if (e.prefix) {
                    this.allowlistedPrefixes =
                        this.allowlistedPrefixes.concat(...e.prefix);
                }
                if (e.regexp) {
                    this.allowlistedRegExps = this.allowlistedRegExps.concat(...e.regexp.map(r => new RegExp(r)));
                }
            }
        }
    }
    isAllowlisted(filePath) {
        if (this.allowlistMemoizer.has(filePath)) {
            return this.allowlistMemoizer.get(filePath);
        }
        for (const p of this.allowlistedPrefixes) {
            if (filePath.startsWith(p)) {
                this.allowlistMemoizer.set(filePath, true);
                return true;
            }
        }
        for (const re of this.allowlistedRegExps) {
            if (re.test(filePath)) {
                this.allowlistMemoizer.set(filePath, true);
                return true;
            }
        }
        this.allowlistMemoizer.set(filePath, false);
        return false;
    }
}
exports.Allowlist = Allowlist;
